{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0a72014-5773-4b26-8a57-a13839dd0654",
   "metadata": {},
   "source": [
    "Source of original data: [https://genomics.senescence.info](https://genomics.senescence.info/longevity/)<br />\n",
    "*(LongevityMap build 3, release date: 2017 June 24, number of genes: 884)*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4b20f4a6-3c4b-4e27-9dd9-bb8636702473",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "94d323f6-6cd9-4a01-a8b1-5c908585918e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (550, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed  Unnamed: 6\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103         NaN\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103         NaN\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664         NaN\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664         NaN\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664         NaN"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read original data into dataFrame\n",
    "df = pd.read_csv('longevity genes -original.csv')\n",
    "\n",
    "print(\"Dimension:\", df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6aa857bc-5345-4a12-9bc2-db500ea3994e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (550, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Last columns of the dataFrame contains only NaN value.\n",
    "# Make sure of this and delete this column as unnecessary.\n",
    "assert df['Unnamed: 6'].isna().all(), \"In the column 'Unnamed: 6' there is at least 1 record with value other than Nan\"\n",
    "del df['Unnamed: 6']\n",
    "\n",
    "print(\"Dimension:\", df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "000ed94e-df60-4d50-abe2-c7c0f181cdcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (16, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>258</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>20824210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>261</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>20824210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>280</td>\n",
       "      <td>significant</td>\n",
       "      <td>European</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>23286790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>281</td>\n",
       "      <td>significant</td>\n",
       "      <td>European</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>23286790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>282</td>\n",
       "      <td>significant</td>\n",
       "      <td>European</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TP53</td>\n",
       "      <td>23286790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id      Association            Population Variant(s) Gene(s)    PubMed\n",
       "255  258      significant  American (Caucasian)        NaN    TP53  20824210\n",
       "256  261  non-significant  American (Caucasian)        NaN    TP53  20824210\n",
       "275  280      significant              European        NaN    TP53  23286790\n",
       "276  281      significant              European        NaN    TP53  23286790\n",
       "277  282      significant              European        NaN    TP53  23286790"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Seperate records that do not have values in the column 'Variant(s)'\n",
    "df_var_nan = df[(df['Variant(s)'].isna())]\n",
    "\n",
    "print(\"Dimension:\", df_var_nan.shape)\n",
    "df_var_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e787cf1-d01d-4844-8ec5-6df5742d3b08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (534, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rows that have some info in the column 'Variant(s)' - this is intermediate dataFrame for further splitting\n",
    "df_var_not_nan = df[(df['Variant(s)'].notna())]\n",
    "\n",
    "print(\"Dimension:\", df_var_not_nan.shape)\n",
    "df_var_not_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8779375f-6c5a-437b-ac08-62679381cd28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (322, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id      Association Population Variant(s)   Gene(s)   PubMed\n",
       "0  1  non-significant      Dutch    HLA-B40     HLA-B  1859103\n",
       "1  2  non-significant      Dutch   HLA-DRB5  HLA-DRB5  1859103\n",
       "2  3  non-significant    Finnish       APOB      APOB  8018664\n",
       "3  4      significant    Finnish      APOC3     APOC3  8018664\n",
       "4  5      significant    Finnish   E2/E3/E4      APOE  8018664"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rows that have sting in the column 'Variant(s)' that begins with characters other than 'rs'\n",
    "df_var_not_rs = df_var_not_nan[~df_var_not_nan['Variant(s)'].str.startswith('rs')]\n",
    "\n",
    "print(\"Dimension:\", df_var_not_rs.shape)\n",
    "df_var_not_rs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2fbb38fb-e969-4f3c-b4e3-6504986bb543",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (212, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>G548</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs1685354,rs647126</td>\n",
       "      <td>UCP3</td>\n",
       "      <td>22743239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>G549</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Jordanian</td>\n",
       "      <td>rs2241766,rs266729</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>20201642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457931,rs1321312,rs4331968,rs9470367,rs6920...</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164,Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population  \\\n",
       "539  G548  non-significant                Danish   \n",
       "540  G549  non-significant             Jordanian   \n",
       "541  G550  non-significant               Italian   \n",
       "543  G552  non-significant                Danish   \n",
       "546   556      significant  American (Caucasian)   \n",
       "\n",
       "                                            Variant(s)              Gene(s)  \\\n",
       "539                                 rs1685354,rs647126                 UCP3   \n",
       "540                                 rs2241766,rs266729               ADIPOQ   \n",
       "541  rs6457931,rs1321312,rs4331968,rs9470367,rs6920...  PANDAR,CDKN1A,RAB44   \n",
       "543                                     rs2866164,Q95H                 MTTP   \n",
       "546                                          rs1042714                ADRB2   \n",
       "\n",
       "       PubMed  \n",
       "539  22743239  \n",
       "540  20201642  \n",
       "541  20126416  \n",
       "543  16015282  \n",
       "546  20399803  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rows that have sting in the column 'Variant(s)' that begins with characters 'rs'\n",
    "df_var_rs = df_var_not_nan[df_var_not_nan['Variant(s)'].str.startswith('rs')]\n",
    "\n",
    "print(\"Dimension:\", df_var_rs.shape)\n",
    "df_var_rs.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a7402eb9-d2ec-4395-a0f9-3875ef1daf29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure that we did not lost something during splitting of the original dataFrame\n",
    "assert len(df) == len(df_var_nan) + len(df_var_rs) + len(df_var_not_rs), \"Something wrong with splitting\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e95c29df-2e00-483b-9971-16072d49f6b2",
   "metadata": {},
   "source": [
    "<br />\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d810269b-c640-411c-b641-d7e407588d6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of list: 212\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[2, 2, 31, 2, 1]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of variants in each record in dataFrame with rs-variants.\n",
    "# It is calculated by the number of commas +1\n",
    "nmb_repeats = (df_var_rs['Variant(s)'].str.count(',') + 1).tolist()\n",
    "\n",
    "assert len(df_var_rs) == len(nmb_repeats), \"Problem with calculation\"\n",
    "print('Length of list:', len(nmb_repeats))\n",
    "nmb_repeats[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0b766727-07c6-4690-b4ac-a89af5be1693",
   "metadata": {},
   "outputs": [],
   "source": [
    "## alternative way:\n",
    "# nmb_repeats_alt = [len(st.split(',')) for st in df_var_rs['Variant(s)']]\n",
    "\n",
    "## make sure that two ways gives the same result\n",
    "# assert len(nmb_repeats_alt) == len(nmb_repeats), \"Problems with calculation\"\n",
    "# assert nmb_repeats_alt == nmb_repeats, \"Problems with calculation (2)\"\n",
    "# nmb_repeats_alt[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "29a661aa-5ca5-4898-81ef-01d87a8772a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3031, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457931,rs1321312,rs4331968,rs9470367,rs6920...</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457931,rs1321312,rs4331968,rs9470367,rs6920...</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164,Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164,Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population  \\\n",
       "541  G550  non-significant               Italian   \n",
       "541  G550  non-significant               Italian   \n",
       "543  G552  non-significant                Danish   \n",
       "543  G552  non-significant                Danish   \n",
       "546   556      significant  American (Caucasian)   \n",
       "\n",
       "                                            Variant(s)              Gene(s)  \\\n",
       "541  rs6457931,rs1321312,rs4331968,rs9470367,rs6920...  PANDAR,CDKN1A,RAB44   \n",
       "541  rs6457931,rs1321312,rs4331968,rs9470367,rs6920...  PANDAR,CDKN1A,RAB44   \n",
       "543                                     rs2866164,Q95H                 MTTP   \n",
       "543                                     rs2866164,Q95H                 MTTP   \n",
       "546                                          rs1042714                ADRB2   \n",
       "\n",
       "       PubMed  \n",
       "541  20126416  \n",
       "541  20126416  \n",
       "543  16015282  \n",
       "543  16015282  \n",
       "546  20399803  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dublicate records in the dataFrame 'df_var_rs' that have several variants in the column 'Variant(s)'\n",
    "df_var_rs_extended = df_var_rs.loc[df_var_rs.index.repeat(nmb_repeats)]\n",
    "\n",
    "print(\"Dimension:\", df_var_rs_extended.shape)\n",
    "df_var_rs_extended.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c72fdf63-a025-4358-a92b-415f00d54b95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<generator object <genexpr> at 0x00000230AD0B0900>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create \"list\" of all variants in the column 'Variant(s)' in the dataFrame 'df_var_rs'.\n",
    "# To speed up calculation, use generator.\n",
    "gen_variants = (el for ls in df_var_rs['Variant(s)'] for el in ls.split(','))\n",
    "gen_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "97e8a8fe-ac1d-414e-82f8-9e49f39147bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Look at result (warning: the generator will be exhausted)\n",
    "# ls_variants = list(gen_variants)\n",
    "# print('Length of resulting list:', len(ls_variants))\n",
    "# ls_variants[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a59b96b8-28c0-4643-aeb6-eb500f01cd6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## alternative way:\n",
    "# import itertools\n",
    "# iter_variants = itertools.chain.from_iterable(ls.split(',') for ls in df_var_rs['Variant(s)'])\n",
    "# print(type(iter_variants))\n",
    "\n",
    "## look at result (warning: the iterator will be exhausted) and make sure that two ways gives the same result\n",
    "# ls_variants_alt = list(iter_variants)\n",
    "\n",
    "# assert len(ls_variants_alt) == len(ls_variants), \"Problem with calculation\"\n",
    "# print('Length of resulting list:', len(ls_variants_alt))\n",
    "\n",
    "# assert ls_variants_alt == ls_variants, \"Problem with calculation (2)\"\n",
    "# ls_variants_alt[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2a493e1d-e4f6-4b3d-9dba-8a0ce3585706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3031, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457940</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs2145047</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association            Population Variant(s)  \\\n",
       "541  G550  non-significant               Italian  rs6457940   \n",
       "541  G550  non-significant               Italian  rs2145047   \n",
       "543  G552  non-significant                Danish  rs2866164   \n",
       "543  G552  non-significant                Danish       Q95H   \n",
       "546   556      significant  American (Caucasian)  rs1042714   \n",
       "\n",
       "                 Gene(s)    PubMed  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541  PANDAR,CDKN1A,RAB44  20126416  \n",
       "543                 MTTP  16015282  \n",
       "543                 MTTP  16015282  \n",
       "546                ADRB2  20399803  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change content of the columns 'Variant(s)' by splitting data\n",
    "df_var_rs_extended['Variant(s)'] = list(gen_variants)\n",
    "\n",
    "print(\"Dimension:\", df_var_rs_extended.shape)\n",
    "df_var_rs_extended.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0e826ea-a7ec-4e6c-a7bb-6bfa55e00325",
   "metadata": {},
   "source": [
    "<br />\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4da7e5a0-02aa-4654-b7d0-1bf3dae8959a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimension: (3369, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Association</th>\n",
       "      <th>Population</th>\n",
       "      <th>Variant(s)</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>PubMed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs9470367</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6920453</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs9462209</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs4713999</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs1321309</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs4711459</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs4714003</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs10947623</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs12192827</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs12192877</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs4135234</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3829963</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3829965</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs4135237</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3829966</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3829967</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs733590</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs762623</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs2395655</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs730506</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs4151702</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3176343</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3176344</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs3176349</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs876581</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457938</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs6457940</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>G550</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>rs2145047</td>\n",
       "      <td>PANDAR,CDKN1A,RAB44</td>\n",
       "      <td>20126416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>rs2866164</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>G552</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>Q95H</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>16015282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>556</td>\n",
       "      <td>significant</td>\n",
       "      <td>American (Caucasian)</td>\n",
       "      <td>rs1042714</td>\n",
       "      <td>ADRB2</td>\n",
       "      <td>20399803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-B40</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Dutch</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>HLA-DRB5</td>\n",
       "      <td>1859103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8018664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>significant</td>\n",
       "      <td>French</td>\n",
       "      <td>I/D</td>\n",
       "      <td>ACE</td>\n",
       "      <td>8136829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>8155090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9</td>\n",
       "      <td>significant</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8155090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10</td>\n",
       "      <td>significant</td>\n",
       "      <td>Swedish</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8403397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>11</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>8541369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>12</td>\n",
       "      <td>significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>APOB</td>\n",
       "      <td>APOB</td>\n",
       "      <td>9050915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>13</td>\n",
       "      <td>significant</td>\n",
       "      <td>English (Cambridge)</td>\n",
       "      <td>I/D</td>\n",
       "      <td>ACE</td>\n",
       "      <td>9105559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14</td>\n",
       "      <td>significant</td>\n",
       "      <td>English (Cambridge)</td>\n",
       "      <td>APOC1</td>\n",
       "      <td>APOC1</td>\n",
       "      <td>9105559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>15</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>English (Cambridge)</td>\n",
       "      <td>E2/E3/E4</td>\n",
       "      <td>APOE</td>\n",
       "      <td>9105559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>16</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>English (Cambridge)</td>\n",
       "      <td>677C/T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>9105559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>17</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>R353Q</td>\n",
       "      <td>F7</td>\n",
       "      <td>9108791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>18</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>-455G/A</td>\n",
       "      <td>FGB</td>\n",
       "      <td>9108791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>19</td>\n",
       "      <td>significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>1-bp guanine deletion/insertion (4G/5G)</td>\n",
       "      <td>SERPINE1</td>\n",
       "      <td>9108791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>20</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Danish</td>\n",
       "      <td>I/D</td>\n",
       "      <td>ACE</td>\n",
       "      <td>9170397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>21</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Spanish</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>9177672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>22</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Spanish</td>\n",
       "      <td>CYP2E1</td>\n",
       "      <td>CYP2E1</td>\n",
       "      <td>9177672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>23</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Spanish</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>9177672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>24</td>\n",
       "      <td>significant</td>\n",
       "      <td>French</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>9425225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>25</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>French</td>\n",
       "      <td>APOA4</td>\n",
       "      <td>APOA4</td>\n",
       "      <td>9533408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>27</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>French</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>9654200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>28</td>\n",
       "      <td>significant</td>\n",
       "      <td>French</td>\n",
       "      <td>I/D</td>\n",
       "      <td>ACE</td>\n",
       "      <td>9761238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>29</td>\n",
       "      <td>significant</td>\n",
       "      <td>English (Caucasian) and Nigerian</td>\n",
       "      <td>638G/A</td>\n",
       "      <td>SULT1A1</td>\n",
       "      <td>9854023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>30</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>9887369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>31</td>\n",
       "      <td>non-significant</td>\n",
       "      <td>Italian</td>\n",
       "      <td>REN</td>\n",
       "      <td>REN</td>\n",
       "      <td>9887369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id      Association                        Population  \\\n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "541  G550  non-significant                           Italian   \n",
       "543  G552  non-significant                            Danish   \n",
       "543  G552  non-significant                            Danish   \n",
       "546   556      significant              American (Caucasian)   \n",
       "0       1  non-significant                             Dutch   \n",
       "1       2  non-significant                             Dutch   \n",
       "2       3  non-significant                           Finnish   \n",
       "3       4      significant                           Finnish   \n",
       "4       5      significant                           Finnish   \n",
       "5       6      significant                            French   \n",
       "6       8      significant                           Finnish   \n",
       "7       9      significant                           Finnish   \n",
       "8      10      significant                           Swedish   \n",
       "9      11  non-significant                           Italian   \n",
       "10     12      significant                           Italian   \n",
       "11     13      significant               English (Cambridge)   \n",
       "12     14      significant               English (Cambridge)   \n",
       "13     15  non-significant               English (Cambridge)   \n",
       "14     16  non-significant               English (Cambridge)   \n",
       "15     17  non-significant                           Italian   \n",
       "16     18  non-significant                           Italian   \n",
       "17     19      significant                           Italian   \n",
       "18     20  non-significant                            Danish   \n",
       "19     21  non-significant                           Spanish   \n",
       "20     22  non-significant                           Spanish   \n",
       "21     23  non-significant                           Spanish   \n",
       "22     24      significant                            French   \n",
       "23     25  non-significant                            French   \n",
       "24     27  non-significant                            French   \n",
       "25     28      significant                            French   \n",
       "26     29      significant  English (Caucasian) and Nigerian   \n",
       "27     30  non-significant                           Italian   \n",
       "28     31  non-significant                           Italian   \n",
       "\n",
       "                                  Variant(s)              Gene(s)    PubMed  \n",
       "541                                rs9470367  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs6920453  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs9462209  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs4713999  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs1321309  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs4711459  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs4714003  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                               rs10947623  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                               rs12192827  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                               rs12192877  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs4135234  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3829963  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3829965  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs4135237  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3829966  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3829967  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                 rs733590  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                 rs762623  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs2395655  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                 rs730506  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs4151702  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3176343  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3176344  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs3176349  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                 rs876581  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs6457938  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs6457940  PANDAR,CDKN1A,RAB44  20126416  \n",
       "541                                rs2145047  PANDAR,CDKN1A,RAB44  20126416  \n",
       "543                                rs2866164                 MTTP  16015282  \n",
       "543                                     Q95H                 MTTP  16015282  \n",
       "546                                rs1042714                ADRB2  20399803  \n",
       "0                                    HLA-B40                HLA-B   1859103  \n",
       "1                                   HLA-DRB5             HLA-DRB5   1859103  \n",
       "2                                       APOB                 APOB   8018664  \n",
       "3                                      APOC3                APOC3   8018664  \n",
       "4                                   E2/E3/E4                 APOE   8018664  \n",
       "5                                        I/D                  ACE   8136829  \n",
       "6                                       APOB                 APOB   8155090  \n",
       "7                                   E2/E3/E4                 APOE   8155090  \n",
       "8                                   E2/E3/E4                 APOE   8403397  \n",
       "9                                   E2/E3/E4                 APOE   8541369  \n",
       "10                                      APOB                 APOB   9050915  \n",
       "11                                       I/D                  ACE   9105559  \n",
       "12                                     APOC1                APOC1   9105559  \n",
       "13                                  E2/E3/E4                 APOE   9105559  \n",
       "14                                    677C/T                MTHFR   9105559  \n",
       "15                                     R353Q                   F7   9108791  \n",
       "16                                   -455G/A                  FGB   9108791  \n",
       "17   1-bp guanine deletion/insertion (4G/5G)             SERPINE1   9108791  \n",
       "18                                       I/D                  ACE   9170397  \n",
       "19                                    CYP2D6               CYP2D6   9177672  \n",
       "20                                    CYP2E1               CYP2E1   9177672  \n",
       "21                                      NAT2                 NAT2   9177672  \n",
       "22                                  HLA-DRB1             HLA-DRB1   9425225  \n",
       "23                                     APOA4                APOA4   9533408  \n",
       "24                                     GSTM1                GSTM1   9654200  \n",
       "25                                       I/D                  ACE   9761238  \n",
       "26                                    638G/A              SULT1A1   9854023  \n",
       "27                                     PARP1                PARP1   9887369  \n",
       "28                                       REN                  REN   9887369  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine all dataFrames in one\n",
    "df_result = pd.concat([df_var_rs_extended, df_var_not_rs, df_var_nan])\n",
    "\n",
    "print(\"Dimension:\", df_result.shape)\n",
    "df_result.iloc[3000:3060]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d0a61b18-d4cb-45b4-876b-50275cc30871",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "# save result to disk\n",
    "df_result.to_csv('longevity genes -splitted.csv', index=False)\n",
    "print('Done')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
